Mage/src/main/java/mage/game/GameImpl.java

```diff
@@ -1451,6 +1452,7 @@ protected void endOfTurn() {
             player.endOfTurn(this);
         }
         state.resetWatchers();
+        state.cleanupPermanentCostsTags(this);
```

Comment:
Need code comments why it clean up on end of the turn and in state's field definition about field lifecycle.